### PR TITLE
Add InitializeInternal concept to Singleton 

### PR DESCRIPTION
**Pull request by keluecke**
Overview
---
This is introducing an abstract initialization step in singleton to help auto sort out dependencies between singletons. Say you have three singletons where singleton A depends on singleton B and uses it for something in it's initialization step. Singleton B, similarly depends on singleton C. If initialization is done in Awake, since unity doesn't guarantee an order in which Awake is called, this could lead to singleton A being initialized first, and throwing errors, similarly with singleton B and C. If you instead follow the practice of initializing in this function, initialization is done on demand, allowing Awake to be called in any order and initialization to be auto sorted, as long as there are no cycles in your dependency graph. It also allows for lazy initialization of singletons.

### CODE REVIEW

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 1]
```diff
@@ -62,11 +72,34 @@ public static bool IsInitialized
             }
         }
 
+        [SerializeField]
+        private bool DontDestroyParentRootOnLoad = true;
+
+        private object initilizedLock = new object();
```
Comment (timGerken) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: spelling

**Comment by timGerken** [comment]
Can you add an example of how this is actually used IRL (for instance the hypothetical you cite in the commit description)? Right now, all Singleton derived classes just have the same empty implementation of:
```
protected override void InitializeInternal()
{

}
```

**Comment by keluecke** [comment]
The example is in this pr: https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/1506

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs.meta [hunk 2]
```diff
@@ -1,8 +1,9 @@
 fileFormatVersion: 2
-guid: b0ed29a95096a25448535f8df9cf0423
-timeCreated: 1455735878
+guid: 3ac3ff05b2d583947859298e1ee17955
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs.meta]: We should probably keep the Singleton meta data the same

**Comment by StephenHodgson** [comment]
Do we need to go though each Singleton and update its `InitializeInternal` implementations?

**Comment by StephenHodgson** [comment]
Where is the example in your PR?  I couldn't find it with all the other changes

**Comment by keluecke** [comment]
Each one can remain empty.  Preferably, initialization happens there, but it doesn't _need_ to.

**Comment by keveleigh** [comment]
There are definitely some places we initialize things in `Awake` or `Start` and have run into initialization order issues where this would be very helpful. The `InputManager`'s event data creation is one that comes to mind.

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs [hunk 3]
```diff
@@ -20,6 +20,10 @@ public class LineManager : Singleton<LineManager>, IGeometry
 
         private const float defaultLineScale = 0.005f;
 
+        protected override void InitializeInternal()
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs]: I'm not a big fan that we have this empty override method in each of the classes.

Is there a way to have the option of using this, without having to have an empty stub in every class that inherits from `Singleton`?

Comment (keluecke) [reply Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs]: We could, but InitializeInternal is meant to solve a problem with script access ordering, which it will only do if it is used to initialize the object.  Having it be optional to include would render it useless, as certainly nobody would use it.

I agree that it would be better to have had this from the start so they are used, and not empty.

Comment (StephenHodgson) [reply Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs]: Idk, the script runtime execution order issues seem to be a one off for most classes, not all.

Option is better in this case, cause it's not something people will have to use every time.

Comment (StephenHodgson) [reply Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs]: Simply making the abstract method virtual in the base Singleton Class should be enough.

Comment (StephenHodgson) [reply Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs]: >Having it be optional to include would render it useless, as certainly nobody would use it.

I disagree, if there's a need, it'll get used.  We just need to make sure we document it correctly.

File: Assets/HoloToolkit/Input/Scripts/Utilities/Managers/MixedRealityCameraManager.cs [hunk 4]
```diff
@@ -60,6 +60,10 @@ public enum DisplayType
         /// </summary>
         public event DisplayEventHandler OnDisplayDetected;
 
+        protected override void InitializeInternal()
```
Comment (NeerajW) [inline Assets/HoloToolkit/Input/Scripts/Utilities/Managers/MixedRealityCameraManager.cs]: Perhaps we use these core scripts to demonstrate the auto resolve.
For example: For MixedRealityCameraManager.cs we must have CameraCache singleton resolved first.

File: Assets/HoloToolkit-Examples/Sharing/SharingService/Scripts/RemoteHeadManager.cs [hunk 5]
```diff
@@ -26,6 +26,10 @@ public class RemoteHeadInfo
         /// </summary>
         private Dictionary<long, RemoteHeadInfo> remoteHeads = new Dictionary<long, RemoteHeadInfo>();
 
+        protected override void InitializeInternal()
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/Sharing/SharingService/Scripts/RemoteHeadManager.cs]: Perhaps we use these core scripts to demonstrate the auto resolve.
For example: For RemoteHeadManager.cs we must have SharingStage singleton resolved first.

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 6]
```diff
@@ -26,20 +27,29 @@ public static T Instance
...
-                    }
+                    return instance;
+                }
+
+                if (!searchForInstance)
+                {
+                    return null;
+                }
+
+                // likly first time
```
Comment (NeerajW) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: Typo: Likely -> Likely

File: Assets/HoloToolkit-Examples/Sharing/SharingService/Scripts/RemoteHeadManager.cs (hunk 5 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit-Examples/Sharing/SharingService/Scripts/RemoteHeadManager.cs]: I agree with this.

Something like `SharingStage.AssertIsInitialized();` Should be enough.

Comment (keluecke) [reply Assets/HoloToolkit-Examples/Sharing/SharingService/Scripts/RemoteHeadManager.cs]: Looking at this I think we want something a little more then that.  If there is no other reason to have RemoteHeadManager initialize on start then ordering, we should be able to move it's initialization to Initialize internal, as well as SharingStage and just have them both call base.Awake().

File: Assets/HoloToolkit/Input/Scripts/Utilities/Managers/MixedRealityCameraManager.cs (hunk 4 above)
Comment (keluecke) [reply Assets/HoloToolkit/Input/Scripts/Utilities/Managers/MixedRealityCameraManager.cs]: CameraCache isn't a singleton method.

File: Assets/HoloToolkit-Examples/Sharing/SharingService/Scripts/RemoteHeadManager.cs (hunk 5 above)
Comment (keluecke) [reply Assets/HoloToolkit-Examples/Sharing/SharingService/Scripts/RemoteHeadManager.cs]: I've run simple tests in the editor for these changes and it runs with each of the scenes in C:\Git\keluecke-holotoolkit-unity\Assets\HoloToolkit-Examples\Sharing\SharingService\Scenes

File: Assets/HoloToolkit/Input/Scripts/Utilities/Managers/MixedRealityCameraManager.cs (hunk 4 above)
Comment (NeerajW) [reply Assets/HoloToolkit/Input/Scripts/Utilities/Managers/MixedRealityCameraManager.cs]: I still don't see a single place where this is getting used in the core toolkit. What about implementing this function in FocusManager.cs?

Comment (keluecke) [reply Assets/HoloToolkit/Input/Scripts/Utilities/Managers/MixedRealityCameraManager.cs]: I added it to Assets/HoloToolkit/Sharing/Scripts/SharingStage.cs so that it is ensured to be initialized when Assets/HoloToolkit-Examples/Sharing/SharingService/Scripts/RemoteHeadManager.cs  calls it.  Would you like it in FocusManager as well?

**Comment by NeerajW** [comment]
After seeing some implementation examples, spoke with Kevin offline. 
I believe PR code reviewers should suggest this feature to new singletons where their is complex initialization logic and not force every singleton by default to use it.

There is a lot of value in adding this change so even if the default is not this we will still benefit from having this functionality. Thank you Kevin for the due diligence.

We agreed on making this optional using override.

**Comment by keluecke** [comment]
Made InitializeInternal virtual instead of abstract.

**Comment by NeerajW** [comment]
@keluecke there are merge conflicts :(

File: Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs [hunk 7]
```diff
@@ -47,26 +44,26 @@ protected override void Awake()
                     RegisterPointer(pointingSource);
                 }
             }
-        }
 
-        private void Start()
-        {
-            if (pointers.Count == 0 && autoRegisterGazePointerIfNoPointersRegistered && GazeManager.IsInitialized)
+            if (pointers.Count == 0 && autoRegisterGazePointerIfNoPointersRegistered && GazeManager.CofirmInitialized())
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs]: I change a good deal of things here in the focus manager concerning pointers. Probably not a good use case example.

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 8]
```diff
@@ -85,9 +132,8 @@ protected virtual void Awake()
             }
             else if (!IsInitialized)
             {
-                instance = (T)this;
+                Initialize();
                 searchForInstance = false;
-                gameObject.GetParentRoot().DontDestroyOnLoad();
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: Why are we removing this?

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 9]
```diff
@@ -62,11 +72,48 @@ public static bool IsInitialized
...
+        {
+            lock (initializedLock)
+            {
+                if (!IsInitialized)
+                {
+                    instance = (T)this;
+                    InitializeInternal();
+                    if (DontDestroyParentRootOnLoad)
+                    {
+                        instance.gameObject.GetParentRoot().DontDestroyOnLoad();;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: Oh I see it was moved here

**Comment by StephenHodgson** [comment]
@NeerajW I think more discussion on this is needed.

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 10]
```diff
@@ -62,11 +72,48 @@ public static bool IsInitialized
             }
         }
 
+        public static bool CofirmInitialized()
```
Comment (keveleigh) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: typo: Cofirm -> Confirm

**Comment by StephenHodgson** [comment]
https://github.com/Microsoft/MixedRealityToolkit-Unity/issues/1654#issuecomment-362020687

>We can do this for all singleton classes that don't need to to inherit from Monobehavior but they'll need to be static.
```
[RuntimeInitializeOnLoadMethod(loadType: RuntimeInitializeLoadType.BeforeSceneLoad)]
private static void RunInitializeInPlayer()
{
    // We're using this method just to make sure the class constructor is called
    // so we don't need any code in here. When the engine calls this method, the
    // class constructor will be run if it hasn't been run already.
}
```
>This guarantees we have an instance before any other Unity API is called. (Awake, OnEnable, Start, etc)

**Comment by keluecke** [comment]
@StephenHodgson @NeerajW I have addressed all the issues, can we move forward, this is dragging on quite long...

File: Assets/HoloToolkit-Examples/Sharing/SharingService/Scripts/RemoteHeadManager.cs [hunk 11]
```diff
@@ -41,6 +41,12 @@ private void Start()
             }
         }
 
+        protected override void Awake()
+        {
+            // We don't want lazy evaluational for this class 
+            base.Awake();
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/Sharing/SharingService/Scripts/RemoteHeadManager.cs]: We can probably remove this method. Redundant anyway.

File: Assets/HoloToolkit-Examples/SpatialMapping/Scripts/PlaneFindingTest.cs [hunk 12]
```diff
@@ -111,7 +111,7 @@ private void OnDrawGizmosSelected()
                     UnityEditor.Handles.DrawLine(corners[1], corners[2]);
                     UnityEditor.Handles.DrawLine(corners[1], corners[3]);
                     UnityEditor.Handles.DrawLine(corners[2], corners[3]);
-                    UnityEditor.Handles.ArrowHandleCap(0, center, Quaternion.FromToRotation(Vector3.forward, normal), 0.4f, EventType.ignore);
+                    UnityEditor.Handles.ArrowHandleCap(0, center, Quaternion.FromToRotation(Vector3.forward, normal), 0.4f, EventType.Ignore);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/SpatialMapping/Scripts/PlaneFindingTest.cs]: Should probably revert this change. We're not using the 2017.3 Editor for the dev branch.

File: Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs [hunk 13]
```diff
@@ -47,26 +44,26 @@ protected override void Awake()
...
-            if (pointers.Count == 0 && autoRegisterGazePointerIfNoPointersRegistered && GazeManager.IsInitialized)
+            if (pointers.Count == 0 && autoRegisterGazePointerIfNoPointersRegistered && GazeManager.ConfirmInitialized())
             {
                 RegisterPointer(GazeManager.Instance);
             }
         }
 
+        protected override void Awake()
+        {
+            base.Awake();
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs]: Remove redundant method call

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs [hunk 14]
```diff
@@ -6,12 +6,13 @@
...
 {
     /// <summary>
-    /// Singleton behaviour class, used for components that should only have one instance.
+    /// Singleton behavior class, used for components that should only have one instance.
     /// <remarks>Singleton classes live on through scene transitions and will mark their 
     /// parent root GameObject with <see cref="Object.DontDestroyOnLoad"/></remarks>
     /// </summary>
     /// <typeparam name="T">The Singleton Type</typeparam>
-    public class Singleton<T> : MonoBehaviour where T : Singleton<T>
+    [DisallowMultipleComponent]
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Common/Scripts/Singleton.cs]: Great addition!

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs [hunk 15]
```diff
@@ -6,12 +6,13 @@
 namespace HoloToolkit.Unity
 {
     /// <summary>
-    /// Singleton behaviour class, used for components that should only have one instance.
+    /// Singleton behavior class, used for components that should only have one instance.
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Common/Scripts/Singleton.cs]: Why rename this?
Unsure but I think the spelling is correct here.

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs.meta [hunk 16]
```diff
@@ -1,8 +1,9 @@
 fileFormatVersion: 2
-guid: b0ed29a95096a25448535f8df9cf0423
-timeCreated: 1455735878
+guid: 3ac3ff05b2d583947859298e1ee17955
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Common/Scripts/Singleton.cs.meta]: please revert this file. May have unintended consequences in other people's projects.

File: Assets/HoloToolkit/Common/Scripts/WorldAnchorManager.cs [hunk 17]
```diff
@@ -89,7 +89,7 @@ protected enum AnchorOperation
         /// Internal list of anchors and their GameObject references.
         /// </summary>
         protected Dictionary<string, GameObject> AnchorGameObjectReferenceList = new Dictionary<string, GameObject>(0);
-
+
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Common/Scripts/WorldAnchorManager.cs]: nit: revert

File: Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs [hunk 18]
```diff
@@ -153,6 +151,11 @@ protected override void Awake()
             FindGazeTransform();
         }
 
+        protected override void Awake()
+        {
+            base.Awake();
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs]: nit: remove redundant base call and and method.

File: Assets/HoloToolkit/Sharing/Scripts/SharingStage.cs [hunk 19]
```diff
@@ -239,6 +235,14 @@ protected override void Awake()
             }
         }
 
+        #region Unity Methods
+
+        protected override void Awake()
+        {
+            // We don't want lazy initialization for this class
+            base.Awake();
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Sharing/Scripts/SharingStage.cs]: nit: remove redundant base call and and method.

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs (hunk 15 above)
Comment (keluecke) [reply Assets/HoloToolkit/Common/Scripts/Singleton.cs]: with the u has to be a British English version, I think bot hare correct enough, my comment spell checker just flagged it.

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs [hunk 20]
```diff
@@ -62,11 +72,48 @@ public static bool IsInitialized
...
+        {
+            lock (initializedLock)
+            {
+                if (!IsInitialized)
+                {
+                    instance = (T)this;
+                    InitializeInternal();
+                    if (DontDestroyParentRootOnLoad)
+                    {
+                        instance.gameObject.GetParentRoot().DontDestroyOnLoad();;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Common/Scripts/Singleton.cs]: We should probably use `DontDestroyOnLoad(instance.transform.root);`

Also, remove extra `;`

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs [hunk 21]
```diff
@@ -62,11 +72,48 @@ public static bool IsInitialized
             }
         }
 
+        public static bool ConfirmInitialized()
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Common/Scripts/Singleton.cs]: This method seems rather redundant to `AssertIsInitialized()` and `IsInitialized` property.

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs (hunk 14 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit/Common/Scripts/Singleton.cs]: I tried this out but it doesn't actually seem to do anything in the editor.

I was able to add a gaze manager component multiple times anyway.

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs [hunk 22]
```diff
@@ -62,11 +72,48 @@ public static bool IsInitialized
...
+        }
+
+        [SerializeField]
+        private bool DontDestroyParentRootOnLoad = true;
+
+        private object initializedLock = new object();
+
+        protected void Initialize()
+        {
+            lock (initializedLock)
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Common/Scripts/Singleton.cs]: What's the purpose of the lock here?

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RemoveSurfaceVertices.cs [hunk 23]
```diff
@@ -52,7 +52,6 @@ public class RemoveSurfaceVertices : Singleton<RemoveSurfaceVertices>
         private static readonly float FrameTime = .008f;
 #endif
 
-        // GameObject initialization.
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RemoveSurfaceVertices.cs]: nit: revert. Does not relate to PR

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/SurfaceMeshesToPlanes.cs [hunk 24]
```diff
@@ -92,7 +92,6 @@ public class SurfaceMeshesToPlanes : Singleton<SurfaceMeshesToPlanes>
         private static readonly float FrameTime = .008f;
 #endif
 
-        // GameObject initialization.
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/SurfaceMeshesToPlanes.cs]: nit: revert. Does not relate to PR

File: Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs [hunk 25]
```diff
@@ -62,11 +56,9 @@ private void Update()
             UpdatePointers();
             UpdateFocusedObjects();
         }
-
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs]: Why remove these spaces?

**Comment by StephenHodgson** [comment]
Not really sure this PR is needed if we just [update the Script Execution Order](https://github.com/Microsoft/MixedRealityToolkit-Unity/issues/1654#issuecomment-362392932) of the scripts as @Railboy pointed out.

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs [hunk 26]
```diff
@@ -26,20 +27,29 @@ public static T Instance
...
+
+                if (!searchForInstance)
+                {
+                    return null;
+                }
+
+                // likely first time
+                T[] objects = FindObjectsOfType<T>();
+                searchForInstance = false;
+
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Common/Scripts/Singleton.cs]: We could check to see if we need to instantiate a new instance here by searching for or creating a new `GameObject` to attach this component to.

For example:
```
if(objects.Length==0)
{
    instance = new GameObject(typeof(T).Name).AddComponent<T>();
}
```

**Comment by keluecke** [comment]
Meta file changes are a much less discoverable and robust way of doing this.

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs (hunk 22 above)
Comment (keluecke) [reply Assets/HoloToolkit/Common/Scripts/Singleton.cs]: In case an object is accessed on the main thread and another thread at the same time.  Definitely not common but since we have it might as well keep the protection.

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs (hunk 21 above)
Comment (keluecke) [reply Assets/HoloToolkit/Common/Scripts/Singleton.cs]: It's more useful then either of them.

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs (hunk 26 above)
Comment (keluecke) [reply Assets/HoloToolkit/Common/Scripts/Singleton.cs]: You are welcome to add new functionality in a new PR, that is out of scope of this PR.

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs (hunk 21 above)
Comment (keveleigh) [reply Assets/HoloToolkit/Common/Scripts/Singleton.cs]: > It's more useful then either of them.

Why? I'm not familiar with what `T access = Instance;` is doing.

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs (hunk 15 above)
Comment (keveleigh) [reply Assets/HoloToolkit/Common/Scripts/Singleton.cs]: It should probably have the u, to match the spelling of MonoBehaviour.

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs (hunk 21 above)
Comment (keluecke) [reply Assets/HoloToolkit/Common/Scripts/Singleton.cs]: Only when you attempt to access the singleton does it attempt to initialize it.  AssertIsInitialized just throws an exception if it hasn't happened to load yet, even if it would once accessed.  This actually tries to Initialize it then tells you if there is one.   Similarly, IsInitialized doesn't tell you if it would be initialized once you try to use it.

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs (hunk 14 above)
Comment (keluecke) [reply Assets/HoloToolkit/Common/Scripts/Singleton.cs]: Sounds like a bug report for unity

File: Assets/HoloToolkit/Common/Scripts/Singleton.cs (hunk 15 above)
Comment (keluecke) [reply Assets/HoloToolkit/Common/Scripts/Singleton.cs]: fair enough

**Comment by keluecke** [comment]
Thanks all!

**Comment by DDReaper** [comment]
This change may become redundant with the move away from Singletons #1654 , but we should ensure this use case is handled in what comes next.

Personally (after doing a full review), Singletons have been grossly overused and in may cases, just because it's in the scene and needs to be referenced.  Some should just be static, others should become "Managers" in the way forward.
